### PR TITLE
Sandcastle ref mapper

### PR DESCRIPTION
**Pull request by SamuelEnglard**
Implementation for #4069

### CODE REVIEW

**Comment by superyyrrzz** [comment]
Thank you for the PR! I am busy today and will find some time to look into this.

File: tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net472</TargetFramework>
```
Comment (superyyrrzz) [inline tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj]: Can you change it to net 462? Just to align with other projects currently

File: tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    <Copyright>© Microsoft Corporation. All rights reserved.</Copyright>
+    <Product>SandcastleRefMapper</Product>
+    <Company>Microsoft</Company>
+    <Authors>Microsoft</Authors>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.DocAsCode.Metadata.ManagedReference" Version="2.40.12" />
```
Comment (superyyrrzz) [inline tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj]: Change `Microsoft.DocAsCode.Metadata.ManagedReference` and `Microsoft.DocAsCode.Build.Engine` from `PackageReference` to `ProjectReference`.

File: tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net472</TargetFramework>
+    <Copyright>© Microsoft Corporation. All rights reserved.</Copyright>
+    <Product>SandcastleRefMapper</Product>
+    <Company>Microsoft</Company>
+    <Authors>Microsoft</Authors>
```
Comment (superyyrrzz) [inline tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj]: Add ` <LangVersion>latest</LangVersion>` to support async main.

**Comment by superyyrrzz** [comment]
Could you provide some guides about how to use this tool? I have some tests but seems not to make it work. I am not sure whether I pass the correct parameters.

**Comment by superyyrrzz** [comment]
@SamuelEnglard Thank you for the updates. Could you also share some sample usage? That will help understand the code.

**Comment by SamuelEnglard** [comment]
There were some bugs, which I've now fixed. An example though would be `.\DocFx.SandcastleRefMapper.exe --baseUrl https://www.newtonsoft.com/json/help/html/ --outputYamlPath .\json.yml .\Newtonsoft.Json.dll`

File: tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (superyyrrzz) [inline tools/SandcastleRefMapper/DocFx.SandcastleRefMapper.csproj]: Could you remove prefix `DocFx` in the project name, to keep like other tools?

File: tools/SandcastleRefMapper/Program.cs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.DocAsCode.Build.Engine;
+using Microsoft.DocAsCode.Common;
+using Microsoft.DocAsCode.DataContracts.ManagedReference;
+using Microsoft.DocAsCode.Metadata.ManagedReference;
+using Microsoft.DocAsCode.Plugins;
+using CS = Microsoft.CodeAnalysis.CSharp;
+
+namespace DocFx.SandcastleRefMapper
```
Comment (superyyrrzz) [inline tools/SandcastleRefMapper/Program.cs]: `namespace Microsoft.DocAsCode.Tools.SandcastleRefMapper`

**Comment by superyyrrzz** [comment]
@SamuelEnglard Thank you for the update. It runs perfectly on my machine now!

It need some minor work to make it available in our release model, could also add:
* a `.nuspec` under `src/nuspec`
* update `$packages` in `build.ps1`

**Comment by superyyrrzz** [comment]
Hi @SamuelEnglard , any update no this?

**Comment by SamuelEnglard** [comment]
Hi @superyyrrzz, sorry for the late reply. Bunch of stuff came up. This should fix all the issues you found

File: src/nuspec/SandcastleRefMapper/SandcastleRefMapper.nuspec [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  <metadata>
+    <id>SandcastleRefMapper</id>
+    <version>1.0.0</version>
+    <title>SandcastleRefMapper</title>
+    <authors>Microsoft</authors>
+    <owners>Microsoft</owners>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+    <licenseUrl>https://github.com/dotnet/docfx/blob/master/LICENSE</licenseUrl>
+    <projectUrl>https://github.com/dotnet/docfx</projectUrl>
+    <description>This is a DocFX tool to convert Sandcastle documentation to refmaps.</description>
```
Comment (superyyrrzz) [inline src/nuspec/SandcastleRefMapper/SandcastleRefMapper.nuspec]: Some minor adjustment: "This is a DocFX tool to generate cross reference map for Sandcastle documentation site.

**Comment by superyyrrzz** [comment]
Thank you for contribution!

**Pull request merged**